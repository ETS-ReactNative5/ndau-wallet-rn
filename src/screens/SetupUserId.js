import React, { Component } from 'react';
import { View, ScrollView, Text, TextInput, Alert } from 'react-native';
import { SafeAreaView } from 'react-navigation';
import ndauDashboardApi from '../api/NdauDashboardAPI';
import CommonButton from '../components/CommonButton';
import Stepper from '../components/Stepper';
import RNExitApp from 'react-native-exit-app';
import cssStyles from '../css/styles';
import AsyncStorageHelper from '../model/AsyncStorageHelper';
import SetupStore from '../model/SetupStore';

class SetupUserId extends Component {
  constructor(props) {
    super(props);
    this.state = {
      userId: '',
      numberOfAccounts: 0,
      userIdPresent: false
    };
  }

  confirmUserIdPresent = () => {
    if (!this.state.userId) return false;
    return new Promise((resolve, reject) => {
      ndauDashboardApi
        .getNumberOfAccounts(this.state.userId)
        .then((numberOfAccounts) => {
          this.setState({
            numberOfAccounts: numberOfAccounts
          });
          resolve(numberOfAccounts > 0);
        })
        .catch((error) => {
          console.error(error);
          reject(false);
        });
    });
  };

  async verify() {
    if (!this.state.userId) this.showErrorMessage('Please enter a User ID first.');
    const userIdPresent = await this.confirmUserIdPresent();
    this.setState({ userIdPresent: userIdPresent });
    if (!userIdPresent) this.showErrorMessage('No user associated with that ID.');
    if (userIdPresent) this.showInfoMessage('Account verified.');
  }

  showNextSetup = () => {
    SetupStore.setUserId(this.state.userId);
    SetupStore.setNumberOfAccounts(this.state.numberOfAccounts);

    this.props.navigation.navigate('SetupQRCode');
  };

  showExitApp() {
    Alert.alert(
      '',
      `We are thrilled you're excited about ndau! At the moment, this wallet app is only useful to accredited investors who have already purchased ndau. Once you have purchased ndau and received your ID, please come back and set your wallet up.`,
      [
        {
          text: 'Exit app',
          onPress: () => {
            RNExitApp.exitApp();
          }
        }
      ],
      { cancelable: false }
    );
  }

  showErrorMessage(msg) {
    Alert.alert('Error', msg, [ { text: 'OK', onPress: () => {} } ], { cancelable: false });
  }

  showInfoMessage(msg) {
    Alert.alert('Information', msg, [ { text: 'OK', onPress: () => {} } ], { cancelable: false });
  }

  textChanged = (userId) => {
    if (userId.length === 3) {
      userId += '-';
    }
    this.setState({ userId });
  };

  checkIfAlreadyExists = async () => {
    try {
      const userIdAlreadyAdded = await AsyncStorageHelper.doesKeyExist(this.state.userId);
      if (userIdAlreadyAdded) {
        Alert.alert(
          'Warning',
          `You are about to create a NEW recovery phrase for an ID which already has one. Are you sure you want to do this? Your old phrase and addresses will be invalidated.`,
          [
            {
              text: 'Cancel',
              onPress: () => {}
            },
            {
              text: 'Create new',
              onPress: () => {
                this.onSendEmail();
              }
            }
          ],
          { cancelable: false }
        );
      } else {
        this.onSendEmail();
      }
    } catch (error) {
      console.error(error);
      throw error;
    }
  };

  onSendEmail() {
    if (!this.state.userId) this.showErrorMessage('Please enter a User ID first.');
    ndauDashboardApi
      .triggerQRTEmail(this.state.userId)
      .then(() => {
        this.showNextSetup();
      })
      .catch((error) => {
        this.showErrorMessage('Email could not be sent.');
        console.error(error);
      });
  }

  textChanged = (userId) => {
    let u;
    if (userId.substr(-1) === '-') {
      u = userId;
    } else {
      // c = cleaned
      const c = userId.toUpperCase().replace(/[^A-Z0-9]/g, '').split('');
      u = c.length > 3 ? c.map((c, i) => (i === 2 ? `${c}-` : c)).join('') : c.join('');
    }
    this.setState({ userId: u });
  };

  render() {
    return (
      <SafeAreaView style={cssStyles.safeContainer}>
        <View style={cssStyles.container}>
          <ScrollView style={cssStyles.contentContainer}>
            <Stepper screenNumber={1} />
            <View>
              <Text style={cssStyles.wizardText}>
                In order to deliver your ndau to this wallet on Genesis Day, we need the
                six-character code you use to access the ndau dashboard.
              </Text>
            </View>
            <View style={cssStyles.buttonContainer}>
              <TextInput
                style={cssStyles.textInput}
                onChangeText={(userId) => this.textChanged(userId)}
                value={this.state.userId}
                placeholder="Enter your unique User ID"
                placeholderTextColor="#333"
                autoCapitalize="characters"
                maxLength={7}
              />
            </View>

            <View style={cssStyles.buttonContainer}>
              <CommonButton
                onPress={() => {
                  this.verify();
                }}
                title="Verify"
              />
            </View>
            <View style={cssStyles.buttonContainer}>
              <CommonButton onPress={this.showExitApp} title="I don't have an ID" />
            </View>
            <Text style={cssStyles.wizardText}>
              We will send you an email to confirm you are the account holder.
            </Text>
          </ScrollView>
          <View style={cssStyles.footer}>
            <CommonButton
              onPress={() => {
                this.checkIfAlreadyExists();
              }}
              title="Send email"
              disabled={this.state.numberOfAccounts === 0}
            />
          </View>
        </View>
      </SafeAreaView>
    );
  }
}

export default SetupUserId;
