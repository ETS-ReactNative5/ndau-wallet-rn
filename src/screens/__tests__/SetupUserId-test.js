import { mount } from 'enzyme';
import React from 'react';
import { StyleSheet } from 'react-native';
import SetupUserId from '../SetupUserId';

import renderer from 'react-test-renderer';

describe('testing SetupUserId...', () => {
  let styles = StyleSheet.create({
    wizardText: {
      color: '#ffffff',
      fontSize: 20
    }
  });

  beforeEach(() => {});
  it('renders correctly', () => {
    const tree = renderer.create(<SetupUserId parentStyles={styles} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('throws an error if click without userId', () => {
    const wrapper = mount(<SetupUserId parentStyles={styles} />);
    const onlyButton = wrapper.find('Button').at(0);
    onlyButton.simulate('click');
    const alert = wrapper.find('Alert').at(0);
    expect(alert).toBeDefined();
  });

  it('throws an error if click without userId', () => {
    const wrapper = mount(<SetupUserId parentStyles={styles} />);
    const onlyButton = wrapper.find('Button').at(0);
    onlyButton.simulate('click');
    const alert = wrapper.find('Alert').at(0);
    expect(alert).toBeDefined();
  });

  it('does not throw an error if user enters text', () => {
    const wrapper = mount(<SetupUserId parentStyles={styles} />);
    const onlyButton = wrapper.find('Button').at(0);

    wrapper.find('TextInput').at(0).instance().value = 'foo';
    // expect(wrapper.state('userId')).toEqual('foo');

    onlyButton.simulate('click');
    const alert = wrapper.find('Alert').at(0);
    expect(alert);
  });
});
