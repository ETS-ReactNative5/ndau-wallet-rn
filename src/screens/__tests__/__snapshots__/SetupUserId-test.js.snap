// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing SetupUserId... renders correctly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "#1c2227",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1c2227",
        "flex": 1,
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontSize": 20,
              }
            }
          >
            Verify your User ID 
          </Text>
        </View>
        <View>
          <RCTProgressView
            progress={0.125}
            style={
              Array [
                Object {
                  "height": 2,
                },
                undefined,
              ]
            }
          />
        </View>
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontSize": 20,
              }
            }
          >
            In order to deliver your ndau to this wallet on Genesis Day, we need the six-character code you use to access the ndau dashboard.
          </Text>
        </View>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="characters"
          onChangeText={[Function]}
          placeholder="Enter your unique User ID"
          placeholderTextColor="#333"
          value=""
        />
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessible={true}
        focusAnimation={null}
        focusAnimationDuration={225}
        hitSlop={
          Object {
            "bottom": 6,
            "left": 4,
            "right": 4,
            "top": 6,
          }
        }
        onLayout={[Function]}
        onPress={[Function]}
        onPressIn={[Function]}
        onPressOut={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        pointerEvents="box-only"
        rippleCentered={false}
        rippleColor="#0a1724"
        rippleDuration={400}
        rippleFades={true}
        rippleOpacity={0.3}
        rippleSequential={true}
        rippleSize={0}
        style={
          Object {
            "backgroundColor": "#ffffff",
            "borderRadius": 2,
            "height": 36,
            "justifyContent": "space-around",
            "minWidth": 64,
            "paddingHorizontal": 8,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "fontSize": 14,
                "fontWeight": "500",
                "textAlign": "center",
              },
              Object {
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 14,
              },
              Object {
                "color": "#0a1724",
              },
            ]
          }
        >
          VERIFY
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#0a1724",
              "borderRadius": 2,
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <View
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "bottom": 0,
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "borderRadius": 2,
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
